File: lib/to_mdx/parsers/open_document/style/paragraph_properties.rb
# frozen_string_literal: true

require_relative "../base"

module ToMdx
  module Parsers
    module Style
      class ParagraphProperties < Base
        def initialize(element = nil, context = {})
          super(element, context)
        end

        def to_css
          return {} unless @element&.attributes

          css_properties = {}
          @element.attributes.each do |name, value|
            case name
            when "fo:border"
              css_properties["border"] = value
            when "fo:line-height"
              css_properties["line-height"] = value.end_with?("%") ? (value.to_f / 100).to_s : value
            when "fo:margin-bottom"
              css_properties["margin-bottom"] = value unless value.to_f.zero?
            when "fo:margin-left"
              css_properties["margin-left"] = value unless value.to_f.zero?
            when "fo:margin-right"
              css_properties["margin-right"] = value unless value.to_f.zero?
            when "fo:margin-top"
              css_properties["margin-top"] = value unless value.to_f.zero?
            when "fo:text-align"
              css_properties["text-align"] = value unless value == "start"
            when "fo:text-indent"
              css_properties["text-indent"] = value unless value.to_f.zero?
            when "style:writing-mode"
              css_properties["writing-mode"] = value unless value == "lr-tb"
            end
          end

          css_properties
        end

        register_for "style:paragraph-properties"
      end
    end
  end
end

File: src/parsers/open-document/style/paragraph_properties.ts
import { BaseElement } from "../../base_element.ts";

export class ParagraphProperties extends BaseElement {
  toMdx(): string {
    return "";
  }

  toString(): string {
    return "";
  }

  toCss(): Record<string, string> {
    const cssProperties: Record<string, string> = {};

    for (const { name, value } of Array.from(this.element.attributes)) {
      switch (name) {
        case "fo:border":
          cssProperties.border = value;
          break;
        case "fo:line-height":
          if (value.endsWith("%")) {
            // 100% in LibreOffice doesn't look like 1 in CSS, so tweak it a bit to be closer to the original
            // As an example, in Google Slides the font-size was 48px and the translations between lines in the pargraph 58px (≈ 1.2)
            cssProperties["line-height"] = ((parseFloat(value) + 20) / 100).toString();
          } else {
            cssProperties["line-height"] = value;
          }
          break;
        case "fo:margin-bottom":
          if (parseFloat(value) !== 0) {
            cssProperties["margin-bottom"] = value;
          }
          break;
        case "fo:margin-left":
          if (parseFloat(value) !== 0) {
            cssProperties["margin-left"] = value;
          }
          break;
        case "fo:margin-right":
          if (parseFloat(value) !== 0) {
            cssProperties["margin-right"] = value;
          }
          break;
        case "fo:margin-top":
          if (parseFloat(value) !== 0) {
            cssProperties["margin-top"] = value;
          }
          break;
        case "fo:text-align":
          if (value !== "start") {
            cssProperties["text-align"] = value;
          }
          break;
        case "fo:text-indent":
          if (parseFloat(value) !== 0) {
            cssProperties["text-indent"] = value;
          }
          break;
        case "style:writing-mode":
          if (value !== "lr-tb") {
            cssProperties["writing-mode"] = value;
          }
          break;
      }
    }

    return cssProperties;
  }

  static {
    BaseElement.registerFor.call(this, "style:paragraph-properties");
  }
}
